### ISSUE TITLE
OutOfMemoryError instead of Exceptions

### ISSUE THREAD

**Issue by Hakon** [jruby/jruby#216]
Hello, headius suggested i'd file a bug on this.

Sometimes when my code throws an exception instead it sits for over a minute and then throws an OutOfMemoryError.
This example trace is a ActionView::Template::Error but i guess ActionView catches the error and throws its own.

```
ActionView::Template::Error (Java heap space):
  java.nio.HeapCharBuffer.<init>(HeapCharBuffer.java:39)
  java.nio.CharBuffer.allocate(CharBuffer.java:312)
  java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:775)
  java.nio.charset.Charset.decode(Charset.java:771)
  org.jruby.RubyEncoding$UTF8Coder.decode(RubyEncoding.java:246)
  org.jruby.RubyEncoding.decodeUTF8(RubyEncoding.java:192)
  org.jruby.RubyString.decodeString(RubyString.java:719)
  org.jruby.RubyString.toString(RubyString.java:701)
  org.jruby.RubyNameError$RubyNameErrorMessage.to_str(RubyNameError.java:120)
  org.jruby.RubyNameError$RubyNameErrorMessage$INVOKER$i$0$0$to_str.call(RubyNameError$RubyNameErrorMessage$INVOKER$i$0$0$to_str.gen)
  org.jruby.RubyClass.finvoke(RubyClass.java:679)
  org.jruby.javasupport.util.RuntimeHelpers.invoke(RuntimeHelpers.java:541)
  org.jruby.RubyBasicObject.callMethod(RubyBasicObject.java:364)
  org.jruby.util.TypeConverter.convertToType(TypeConverter.java:71)
  org.jruby.util.TypeConverter.convertToType(TypeConverter.java:117)
  org.jruby.RubyBasicObject.convertToString(RubyBasicObject.java:758)
  org.jruby.RubyNameError.to_s(RubyNameError.java:202)
  org.jruby.RubyNameError$INVOKER$i$0$0$to_s.call(RubyNameError$INVOKER$i$0$0$to_s.gen)
  org.jruby.RubyClass.finvoke(RubyClass.java:679)
  org.jruby.javasupport.util.RuntimeHelpers.invoke(RuntimeHelpers.java:541)
  org.jruby.RubyBasicObject.callMethod(RubyBasicObject.java:364)
  org.jruby.RubyException.message(RubyException.java:266)
  org.jruby.RubyException$INVOKER$i$0$0$message.call(RubyException$INVOKER$i$0$0$message.gen)
  org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:292)
  org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:135)
  org.jruby.ast.CallNoArgNode.interpret(CallNoArgNode.java:63)
  org.jruby.evaluator.ASTInterpreter.setupArgs(ASTInterpreter.java:357)
  org.jruby.ast.SuperNode.interpret(SuperNode.java:107)
  org.jruby.ast.NewlineNode.interpret(NewlineNode.java:104)
  org.jruby.ast.BlockNode.interpret(BlockNode.java:71)
  org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:75)
  org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:298)
```

**Comment by mmustala** [comment]
I'm seeing this same issue in my code right now. My code is handling images and other biggish files and it is quite common that when an exception is encountered the process consumes all the heap it has (3 Gigabytes currently) and then the Java heap space is thrown.

I very much suspect the same as headius. What about instead of just to_s  on the object we could take only the first 1000 characters or something.

**Comment by sgonyea** [comment]
@mmustala: I'll drop this in the "ideas bucket." Consider me a crank until @headius weighs in:

Far from an ideal solution, but for objects that can become this large... I'd perhaps suggest monkey patching `#to_s` and using `ObjectSpace.memsize_of(self)` (`require 'objspace'` elsewhere). If it's very, very large then spit out a custom string and perhaps log the source of the issue.

For objects that can become really huge, overriding the behavior of `#inspect` is probably your only reliable course of action. Overriding methods is pretty standard once you start working with abnormally large blobs of data.

**Comment by headius** [comment]
I'm guessing this is actually to blame for the memory explosion on exception raise: https://bugs.ruby-lang.org/issues/9725

Not really sure how to avoid this; it's a weird misfeature of Ruby that it inspects the target of a NameError in this way. Maybe anyone affected by this should +1 the issue I filed.

**Comment by headius** [comment]
MRI has made no movement on the issue I filed, so I feel like we need to fix this unilaterally.

We can mitigate this problem by catching OutOfMemoryError in the NameError logic and falling back on a simple class name. That *could* capture OOME unrelated to the string, but it's very unlikely (and if it's really OOM, another one will come along shortly).

Beyond that, I think we should just modify NameError to not do this. @enebo?

**Comment by JamesIry** [comment]
The problem with OOMError is it's inherently unreliable. If thread A is doing something insane (like building a ginormous string for an exception) and then thread B does a reasonable allocation at the wrong time then B will get an OOM and probably die a horrible death.

I think your discussion on the MRI bug is spot on: you have to prevent the ginormous string allocation in the first place. Detecting the OOM after-the-fact will lead to even harder to diagnose problems, albeit more rarely.

**Comment by sgonyea** [comment]
Hard to say if MRI would accept this change, since it would have dual behavior across Error objects.  It helps that in the land of MRI, memory swap is freeeee! :trollface: 

You should definitely solve this unilaterally.  If it doesn't impact performance too hard, you could even juggle behavior based on the Obj's size...  Or swap in a helpful String to let them know why there's no inspect output.

**Comment by headius** [comment]
There's no bug here for us to fix...it's an unfortunate side effect of standard Ruby behavior.

JRuby 9.1 will not limit memory to 500MB, so perhaps it will rarely run into a NameError that's too big.

If you would like to see this changed, add your vote to the CRuby bug: https://bugs.ruby-lang.org/issues/9725

**Comment by prathamesh-sonpatki** [comment]
@headius Any suggestions for getting around this issue? Looks like we are facing this with JRuby 9.0.40 and Sidekiq.

**Comment by mmustala** [comment]
I didn't find anything else but overriding the `inspect`method in the class that is causing the problem. That way you can make sure inspect is returning a string that is not too big.

**Comment by prathamesh-sonpatki** [comment]
Here is the stacktrace https://gist.github.com/prathamesh-sonpatki/0d6ce40f703a8ce930ff4b008af7002e

**Comment by keithrbennett** [comment]
I'm going to weigh in on this...I don't have any direct experience with this bug but I have been conversing with someone on Stack Overflow who does -- he gave up using JRuby, even though he needed to call Java libraries and it was the obvious choice, because of this problem (see comments in  http://stackoverflow.com/questions/37316055/file-write-happening-asynchronously/37317141?noredirect=1#comment62158497_37317141).

On the higher level, my opinion is that class authors should not presume to know how to_s and inspect are called, and support their use by the (Ruby or Java) VM and frameworks out of their field of influence by ensuring that the size of the strings they return is reasonable (maybe <= 1K?), and that the cost of calling them is low (e.g. not to read an entire DB table). I think we should be able to assume that we can call to_s or inspect to build a log message, for example, and it will not overpower the virtual machine.  If a method that returns a large amount of data is needed, it can be implemented with a new name and called explicitly by those who have the specialized knowledge of it.

So I wonder if we're looking in the wrong place for the solution.  If ActiveRecord includes reads in an entire table and includes its data in a returned string, IMO *this* is the problem.

If we assume that implementers of to_s and inspect should limit the size and cost of the returned strings, then it is reasonable to include them in an exception's text as is currently the case.

However, given the severity of the situation, I think Charlie's suggestion is a quicker solution, and one that is more reliable, since we don't know what classes will be coded in the future.

My vote would be to fix this ASAP, even if it means that JRuby deviates slightly from MRI.

**Comment by enebo** [comment]
@keithrbennett are you sure that was the link you intended to include?  I do not understand how it is related.

As a general statement I find this to be bugs in various libraries who dump the planet.  If you are going to dump a DB as output then perhaps you need to limit output?   As a broader statement it also shows how Ruby might have been better off using IO-based API for this (e.g.  output(print_io=$stdout)).

I have a different proposal for to_s on NameError.  I think arguably this is something where some people will want full output and others will not.  We can make a new require:

```ruby
require 'abbreviated_to_s'

class NameError
   include AbbreviatedToS
end
```

By including this to NameError we will check to see if it is included and then do a respond_to? :abbreviated_to_s and call that instead.

This will put a possible burden on library author to fix the general issue by making their own abbreviated output.   I think though if you plan on not including everything in output it is probably important that library auhors can make a best effort on what is displayed.

A generic implementation can be made as well.  It would need to be somewhat reflective and inspect sizes of various ivars and their values and elide values which are too big.  So this could be something like:

```ruby
require 'abbreviated_to_s/reflective'

class ActiveRecord::Base
   include AbbreviatedToS::Reflective
end
```

Then if people run into this they can add a little magic in a Rails initializer and the problem is fixed. 

So on our side we would need NameError (or anything else which would need this) see if the module in question is in its hierarchy and call it instead of to_s if so.  Someone would need to write these small shims and I think export them as a gem.  

This also feels like it could be baked into JRuby and MRI as well but perhaps just trying to solve this in one impl first might be a good path forward?

Note: abbreviated_to_s and the names I picked are just stakeholder names.  I do not really care what name is used so long as people like a name.

**Comment by headius** [comment]
@tenderlove @sgrif Might either of you know if this sort of problem has been fixed in ActiveRecord? To be more specific: a NameError raised against a model with a lot of relationships might drag along a LOT of output when the NameError is eventually printed out.

See also https://bugs.ruby-lang.org/issues/9725 which has stalled.

**Comment by sgrif** [comment]
I don't believe so, no. I'm not sure if there's much that we can do on the AR side short of re-implementing `NameError`

**Comment by keithrbennett** [comment]
I may be misunderstanding this issue, but...

@sgrif Is it true that ActiveRecord's to_s is pulling in an entire table's data into the string?  If this is so, then I would argue that the issue is not that NameError is calling to_s, but rather that the to_s is building and returning such a huge string to begin with.

I think the higher level question that the community needs to answer is "given the havoc they can wreak, do we want to condone and support huge to_s strings?".

In my opinion, the answer is no, because I think the value of enabling any code to be able to call to_s or inspect on unknown objects without fear of bringing down the system is greater than the value of the freedom to implement huge strings.

@enebo The JRuby discussion was buried in the comments.  Search the page for "JRuby" and you'll see it.  Also, I'm sure your solution is technically sound, you know way more than I about these things, but I still think that's way more complexity than the freedom to return huge strings is worth.

**Comment by sgrif** [comment]
No we don't pull the entire table. On instances we actually shouldn't be grabbing anything from #inspect. On the class object we load the schema if it's not loaded. On a relation instance we will load the query that it represents.

**Comment by headius** [comment]
It seems possible the issues people have had here are not a problem in recent/current Rails...wish we could get some confirmation from someone who has had the issue in the past.

**Comment by sgrif** [comment]
I'd be happy to do more digging if you can point me at more details of what
they were doing that would load so much data

**Comment by mmustala** [comment]
I tried to create a simple app with the 9k jruby to demonstrate this error but couldn't get it reproduced. My original environment where the error happened is not available any more. I may give it another shot later if I find time for it.

**Comment by Hakon** [comment]
I no longer work on the project where I originally experienced the issue in question. I believe the problem was that activerecord added each field of the record in the to_s.
The table in question had quite a big amount of fields and some of the fields contained large amounts of data.

I hope this helps.

**Comment by mrckzgl** [comment]
I hope even without filing a new issue my comment will get attention. I am still facing the OP bug in JRuby 9.4.8.0 and it is essentially making development of our project a pain, since nameerrors might happen very often, and with this bug it is currently impossible to spot them in our code: The OOME do not show the user space location where the NameError is raised, in addition to the large time it takes to collect the inspect string.
I've come up with a simple test case:
```
class Test
  def initialize()
    @obj = {}
    for t in 1..800000 do
      @obj[t.to_s] = Random.rand(10000000).to_s
    end
  end
  def error
    nameerror
  end
end

puts("generating large obj")
test = Test.new
puts("done")
test.error
```
running this with: `jruby -w -J-Xmx250M test2.rb` results in:
```
jruby -w -J-Xmx250M test2.rb
generating large obj
done
Error: Your application used more memory than the safety cap of 250M.
Specify -J-Xmx####M to increase it (#### = cap size in MB).
java.lang.OutOfMemoryError: Java heap space
	at org.jruby.dist/org.jruby.util.ByteList.ensure(ByteList.java:362)
	at org.jruby.dist/org.jruby.util.io.EncodingUtils.strBufCat(EncodingUtils.java:1883)
	at org.jruby.dist/org.jruby.util.io.EncodingUtils.strBufCat(EncodingUtils.java:1868)
	at org.jruby.dist/org.jruby.util.io.EncodingUtils.encCrStrBufCat(EncodingUtils.java:1991)
	at org.jruby.dist/org.jruby.util.io.EncodingUtils.encStrBufCat(EncodingUtils.java:1894)
	at org.jruby.dist/org.jruby.RubyBasicObject.inspectObj(RubyBasicObject.java:1145)
	at org.jruby.dist/org.jruby.RubyBasicObject.hashyInspect(RubyBasicObject.java:1088)
	at org.jruby.dist/org.jruby.RubyBasicObject.inspect(RubyBasicObject.java:1072)
	at org.jruby.dist/org.jruby.RubyKernel.inspect(RubyKernel.java:2362)
	at org.jruby.dist/org.jruby.RubyKernel$INVOKER$s$0$0$inspect.call(RubyKernel$INVOKER$s$0$0$inspect.gen)
	at org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:456)
	at org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:195)
	at org.jruby.dist/org.jruby.RubyBasicObject.rbInspect(RubyBasicObject.java:1097)
	at org.jruby.dist/org.jruby.RubyObject.inspect(RubyObject.java:452)
	at org.jruby.dist/org.jruby.RubyNameError$RubyNameErrorMessage.to_str(RubyNameError.java:129)
	at org.jruby.dist/org.jruby.RubyNameError$RubyNameErrorMessage$INVOKER$i$0$0$to_str.call(RubyNameError$RubyNameErrorMessage$INVOKER$i$0$0$to_str.gen)
	at org.jruby.dist/org.jruby.RubyClass.checkFuncallDefault(RubyClass.java:685)
	at org.jruby.dist/org.jruby.RubyClass.finvokeChecked(RubyClass.java:629)
	at org.jruby.dist/org.jruby.runtime.Helpers.invokeChecked(Helpers.java:757)
	at org.jruby.dist/org.jruby.RubyBasicObject.checkCallMethod(RubyBasicObject.java:341)
	at org.jruby.dist/org.jruby.util.TypeConverter.convertToType(TypeConverter.java:97)
	at org.jruby.dist/org.jruby.util.TypeConverter.convertToType(TypeConverter.java:128)
	at org.jruby.dist/org.jruby.RubyBasicObject.convertToString(RubyBasicObject.java:722)
	at org.jruby.dist/org.jruby.RubyNameError.to_s(RubyNameError.java:264)
	at org.jruby.dist/org.jruby.RubyNameError$INVOKER$i$0$0$to_s.call(RubyNameError$INVOKER$i$0$0$to_s.gen)
	at org.jruby.dist/org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:841)
	at org.jruby.dist/org.jruby.ir.runtime.IRRuntimeHelpers.unresolvedSuper(IRRuntimeHelpers.java:1488)
	at org.jruby.dist/org.jruby.ir.instructions.UnresolvedSuperInstr.interpret(UnresolvedSuperInstr.java:123)
	at org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:363)
	at org.jruby.dist/org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66)
	at org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:76)
	at org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:164)
```

Looking purely on the jruby side, it is caused by:
https://github.com/jruby/jruby/blob/13309db2ae3a792894134604a90b55064ad0fa0f/core/src/main/java/org/jruby/RubyNameError.java#L129
Ironic is that the inspect description is stripped out anyway if larger than 64 characters:
https://github.com/jruby/jruby/blob/13309db2ae3a792894134604a90b55064ad0fa0f/core/src/main/java/org/jruby/RubyNameError.java#L134
...so the larger the inspect string, the more time is wasted creating it for nothing ;-)

@headius there is a comment in the standard repo https://bugs.ruby-lang.org/issues/18285#note-44 which I do not understand fully. Accordingly, https://bugs.ruby-lang.org/issues/9725 , which is essentially the ruby version of this issue should be fixed. But its still open, and the problem is still present in jruby ... How to understand that comment? Will we need to wait for a higher jruby version (using a higher ruby version) for this to get fixed?

Would it be possible, like others suggested above, to fix this unilateral in jruby and just remove the inspect call? This would help alot for our case.

best

**Comment by headius** [comment]
@mrckzgl Thank you for waking this issue up! You found all the right links and this is indeed fixed in standard Ruby version 3.3 and higher, but the current release line of JRuby, 9.4, is 3.1 compatible. I am willing to consider changing this behavior in a minor release of 9.4, but it will definitely be fixed for JRuby 10 due out this January.

@enebo Would it be better to open a new issue or reopen this one? What do you think about making this change in 9.4.x, knowing that it's already standard behavior in CRuby?

**Comment by enebo** [comment]
@headius let's open a new issue. Especially if we plan on this for 9.4. it
will be more clearly documented.

**Comment by headius** [comment]
@mrckzgl could you take your comment above and turn it into a new issue? I think you've described the problem succinctly and provided a good example reproduction. 

The fix will be pretty easy. Hopefully the impact of doing it in 9.4 is extremely minimal.

**Comment by mrckzgl** [comment]
@headius Awesome, thank you very much! Both is very good news, the near release of JRuby 10, and your consideration for a minor fix in 9.4. This helps us alot :-)

I've created the new issue: https://github.com/jruby/jruby/issues/8384
